nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix

```diff
@@ -0,0 +1,215 @@
+    (mkRenamedOptionModule ["services" "hercules-ci-agent" "baseDirectory"] ["services" "hercules-ci-agent" "settings" "baseDirectory"])
+    (mkRenamedOptionModule ["services" "hercules-ci-agent" "concurrentTasks"] ["services" "hercules-ci-agent" "settings" "concurrentTasks"])
+  ];
+
+  options.services.hercules-ci-agent = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Enable to run Hercules CI Agent as a system service.
+
+        <link xlink:href="https://hercules-ci.com">Hercules CI</link> is a
+        continuous integation service that is centered around Nix.
+
+        Support is available at <link xlink:href="mailto:help@hercules-ci.com">help@hercules-ci.com</link>.
+      '';
+    };
+    patchNix = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to set nix.package to a patched pkgs.nix if necessary.
+
+        We strive for minimal patching. This option is only for your convenience,
+        when an important patch is not accepted into the stable branch for
+        example.
+      '';
+    };
+    checkNix = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Whether to make sure that the system's Nix (nix-daemon) is compatible.
+
+        If you set this to false, please keep up with the change log.
+      '';
+    };
+    package = mkOption {
+      description = ''
+        Package containing the bin/hercules-ci-agent executable.
+      '';
+      type = types.package;
+      default = pkgs.hercules-ci-agent;
+      defaultText = literalExample "pkgs.hercules-ci-agent";
+    };
+    settings = mkOption {
+      description = ''
+        These settings are written to the <literal>agent.toml</literal> file.
+
+        Not all settings are listed as options, can be set nonetheless.
+
+        For the exhaustive list of settings, see <link xlink:href="https://docs.hercules-ci.com/hercules-ci/reference/agent-config/">documented on docs.hercules-ci.com</link>.
+      '';
+      type = types.submoduleWith { modules = [ settingsModule ]; };
+    };
+
+    /*
+      Internal and/or computed values.
+
+      These are written as options instead of let binding to allow sharing with
+      default.nix on both NixOS and nix-darwin.
+     */
+    tomlFile = mkOption {
+      type = types.path;
+      internal = true;
+      defaultText = "generated hercules-ci-agent.toml";
+      description = ''
+        The fully assembled config file.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    nix.extraOptions = lib.addContextFrom checkNix ''
+      # A store path that was missing at first may well have finished building,
+      # even shortly after the previous lookup. This *also* applies to the daemon.
+      narinfo-cache-negative-ttl = 0
+    '';
+    nix.package = mkIf cfg.patchNix patchedNix;
```

Comment:
It might be better to use an overlay for this instead, because it will allow users to still specify e.g. `nix.package = pkgs.nixFlakes`, and the overlay can patch all of `pkgs.{nix,nixUnstable,nixFlakes}` (if necessary for each of them).

Reply:
It's only necessary for the daemon and only when its `2.3`. They can set it to `nixFlakes` just fine because that version does not need patching.
It's a temporary situation until `2.4` is released. At that point, I'll drop the whole patching business, so there's little point in improving this.

Reply:
Ah so the intended flow is:
- User enables module
- Because `checkNix = true` by default and `patchNix = false`, user gets error that the Nix check failed and they need to use a version with a patch
- User sets `patchNix = true`, therefore overriding the default Nix version (or I guess it would error if the user also assigned `nix.package` in their config)

I guess that's fine